fuel/src/main/kotlin/com/github/kittinunf/fuel/core/interceptors/ParameterEncoder.kt

```diff
@@ -0,0 +1,79 @@
+package com.github.kittinunf.fuel.core.interceptors
+
+import com.github.kittinunf.fuel.core.FoldableRequestInterceptor
+import com.github.kittinunf.fuel.core.Headers
+import com.github.kittinunf.fuel.core.Method
+import com.github.kittinunf.fuel.core.Parameters
+import com.github.kittinunf.fuel.core.RequestTransformer
+import java.net.URL
+import java.net.URLEncoder
+
+object ParameterEncoder : FoldableRequestInterceptor {
+    override fun invoke(next: RequestTransformer): RequestTransformer {
+        return inner@{ request ->
+            val contentType = request[Headers.CONTENT_TYPE].lastOrNull()
+
+            // Expect the parameters to be already encoded in the body
+            if (contentType?.startsWith("multipart/form-data") == true) {
+                return@inner next(request)
+            }
+
+            // If it can be added to the body
+            if (request.body.isEmpty() && allowParametersInBody(request.method)) {
+                if (contentType.isNullOrBlank() || contentType.startsWith("application/x-www-form-urlencoded")) {
+                    return@inner next(
+                        request
+                            .header(Headers.CONTENT_TYPE, "application/x-www-form-urlencoded")
```

Comment:
[Q] I am confused on this line a bit why do we need to add "application/x-www-form-urlencoded" after the check once again?

Reply:
We want to support the following:

```kotlin
requestWithParams
  .header(Headers.CONTENT_TYPE, "application/x-www-form-urlencoded")
  .await()
```

... but also ...

```
requestWithParams
  .await()
```

Both cases are the same, it needs to encode the parameters into the `body`; but in the second case the header might be `null` or `blank` and still need to be set. We don't want users who opt to set it explicitly to be fucked and have their parameters not encoded 📦

Reply:
AHHH I see what it is now.